internal/github/github.go

```diff
@@ -0,0 +1,140 @@
+// wait-for-github
+// Copyright (C) 2022, Grafana Labs
+
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the GNU Affero General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option) any
+// later version.
+
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+// FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
+// details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package github
+
+import (
+	"context"
+	"fmt"
+
+	ghauth "github.com/beatlabs/github-auth/app"
+	"github.com/beatlabs/github-auth/key"
+	"github.com/google/go-github/v48/github"
+	"github.com/gregjones/httpcache"
+	"golang.org/x/oauth2"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type GithubClient interface {
+	IsPRMergedOrClosed(ctx context.Context, owner, repo string, pr int) (bool, bool, error)
+	GetCIStatus(ctx context.Context, owner, repo string, commitHash string) (CIStatus, error)
+}
+
+type AuthInfo struct {
+	InstallationID string
+	AppID          string
+	PrivateKey     []byte
+
+	GithubToken string
+}
+
+type GHClient struct {
+	client *github.Client
+}
+
+func NewGithubClient(ctx context.Context, authInfo AuthInfo) (GithubClient, error) {
+	// If a GitHub token is provided, use it to authenticate in preference to
+	// App authentication
+	if authInfo.GithubToken != "" {
+		log.Debug("Using GitHub token for authentication")
+		return AuthenticateWithToken(ctx, authInfo.GithubToken), nil
+	}
+
+	// Otherwise, use the App authentication flow
+	log.Debug("Using GitHub App for authentication")
+	return AuthenticateWithApp(ctx, authInfo.PrivateKey, authInfo.AppID, authInfo.InstallationID)
+}
+
+// AuthenticateWithToken authenticates with a GitHub token
+func AuthenticateWithToken(ctx context.Context, token string) GithubClient {
+	src := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	httpClient := oauth2.NewClient(ctx, src)
+	httpClient.Transport = httpcache.NewMemoryCacheTransport()
+	githubClient := github.NewClient(httpClient)
+
+	return &GHClient{client: githubClient}
+}
+
+// AuthenticateWithApp authenticates with a GitHub App
+func AuthenticateWithApp(ctx context.Context, privateKey []byte, appID, installationID string) (GithubClient, error) {
+	// Parse the app's private key
+	privKey, err := key.Parse(privateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create an App Config using the App ID and the private key
+	auth, err := ghauth.NewConfig(appID, privKey)
```

Comment:
Any reason to not use the library recommended by the github client? https://github.com/google/go-github#as-a-github-app
That beatlabs lib is not as popular

Reply:
Not that I know of. I think I found some example of using that with the graphql library I was using at the start. (Can't do the `If-Modified-Since` stuff with GraphQL though so I changed to REST.) I can change to the other one 👍